File: p037.py
import utils

primes = utils.primes(1000000)

def truncatable(n):
    sn = str(n)
    if len(sn) < 2:
        return False

    for i in range(1, len(sn)):
        if int(sn[0:i]) not in primes or int(sn[-i:]) not in primes:
            return False 
    return True

tprimes = filter(truncatable, primes)

print tprimes
print len(tprimes)
print sum(tprimes)

File: go/src/p037/p037.go
package main

import (
    "fmt"
    "strconv"
    "utils"
)

func main() {

    primes := make(map[string]uint64)
    for _, p := range utils.Primes(1000000) {
        primes[strconv.FormatUint(p, 10)] = p
    }

    sum := uint64(0)
    for primeString, prime := range primes {
        if truncatable(primeString, primes) {
            sum += prime
        }
    }

    fmt.Println(sum)
}

func truncatable(p string, primes map[string]uint64) (bool) {
    if len(p) == 1 {
        return false
    }
    for i := 1; i < len(p); i++ {
        if primes[p[i:]] == 0 || primes[p[:i]] == 0 {
            return false
        }
    }
    return true
}
